### PR TITLE
Memory monitoring for jenkins during extended tests

### PR DESCRIPTION
**Pull request by jupierce**
Example entry:
```
Jenkins memory statistics at 2016-11-04 17:28:23.024556251 -0400 EDT
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
1000090+   476  0.0  0.0  47432  3212 ?        Rs   21:28   0:00 ps faux
1000090+     1 30.5  2.7 6063144 438908 ?      Ssl  21:25   0:52 java -Xmx307m -Duser.home=/var/lib/jenkins -Dfile.encoding=UTF8 -jar /usr/lib/jenkins/jenkins.war
cache 90386432
rss 442425344
rss_huge 0
mapped_file 14123008
dirty 20480
writeback 0
swap 180346880
pgpgin 310560
pgpgout 180479
pgfault 194011
pgmajfault 4551
inactive_anon 258682880
active_anon 255033344
inactive_file 4714496
active_file 14323712
unevictable 0
hierarchical_memory_limit 536870912
hierarchical_memsw_limit 9223372036854771712
total_cache 90386432
total_rss 442425344
total_rss_huge 0
total_mapped_file 14123008
total_dirty 20480
total_writeback 0
total_swap 180346880
total_pgpgin 310560
total_pgpgout 180479
total_pgfault 194011
total_pgmajfault 4551
total_inactive_anon 258682880
total_active_anon 255033344
total_inactive_file 4714496
total_active_file 14323712
total_unevictable 0
recent_rotated_anon 15485
recent_rotated_file 520
recent_scanned_anon 26309
recent_scanned_file 719
```

### CODE REVIEW

**Comment by jupierce** [comment]
Attempting to improve diagnostic information available for #11755

**Comment by jupierce** [comment]
@csrwng @bparees @gabemontero

**Comment by bparees** [comment]
looks promising.  let's run the extended tests on this PR a few times before we merge it, see if we can recreate it w/o polluting our normal extended test logs.

**Comment by bparees** [comment]
(I suggest you run the entire ecosystem suite, in case the failures are due to some interactions between tests/overloading the host/etc)

**Comment by jupierce** [comment]
[testextended][extended:core(builds|image_ecosystem)]

**Comment by jupierce** [comment]
[testextended][extended:core(openshift pipeline plugin)]

**Comment by jupierce** [comment]
[testextended][extended:core(image_ecosystem)]

File: test/extended/image_ecosystem/plugin.go [hunk 1]
```diff
@@ -378,6 +378,19 @@ func getAdminPassword(oc *exutil.CLI) string {
 	return "password"
 }
 
+// Finds the pod running Jenkins
+func findJenkinsPod(oc *exutil.CLI) *kapi.Pod {
```
Comment (gabemontero) [inline test/extended/image_ecosystem/plugin.go]: There is some similar logic in test/extended/util/framework.go that searches for pods (DumpDeploymentLogs for example).  Perhaps we should somehow refactor this method with those, making this one more generic and take in args for the prefix / contains calls (maybe even make the phase a param?) and get some reuse.

File: test/extended/image_ecosystem/plugin.go [hunk 2]
```diff
@@ -378,6 +378,19 @@ func getAdminPassword(oc *exutil.CLI) string {
...
+	for _, p := range pods.Items {
+		if strings.HasPrefix(p.Name, "jenkins") && !strings.Contains(p.Name, "deploy") && p.Status.Phase == "Running" {
+			return &p
+		}
+	}
+	g.Fail("Should have detected jenkins pod", 1)
+	return nil
+}
+
 func followDCLogs(oc *exutil.CLI, jenkinsNamespace string) {
```
Comment (gabemontero) [inline test/extended/image_ecosystem/plugin.go]: Yeah I could see this method having usefulness beyond our jenkins tests ... perhaps a candidate as well for moving into test/extended/util as well, with the DC name being a param as well.

**Comment by jupierce** [comment]
From the failing test (test connection test) https://ci.openshift.redhat.com/jenkins/job/test_pr_origin_extended/758/consoleFull

```
Jenkins memory statistics at 2016-11-07 16:03:22.425660015 -0500 EST
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
1000570+   235 19.0  0.0  47420   484 ?        Rs   21:03   0:00 ps faux
1000570+     1  127  3.2 3953156 513644 ?      Ssl  21:01   2:13 java -Xmx307m -Duser.home=/var/lib/jenkins -Dfile.encoding=UTF8 -jar /usr/lib/jenkins/jenkins.war
cache 9945088
rss 526577664
rss_huge 100663296
mapped_file 1089536
swap 0
pgpgin 20574294
pgpgout 20467835
pgfault 931918
pgmajfault 74453
inactive_anon 0
active_anon 526569472
inactive_file 811008
active_file 9019392
unevictable 0
hierarchical_memory_limit 536870912
hierarchical_memsw_limit 9223372036854775807
total_cache 9945088
total_rss 526577664
total_rss_huge 100663296
total_mapped_file 1093632
total_swap 0
total_pgpgin 20574294
total_pgpgout 20467835
total_pgfault 931922
total_pgmajfault 74453
total_inactive_anon 0
total_active_anon 526569472
total_inactive_file 811008
total_active_file 9019392
total_unevictable 0
```

rss(502.18M) + cache(9.48M) = 511.66M  in the last report before the test failed.

Interestingly, the JVM RSS value is significantly higher than steady-state Jenkins running in my development environment. If I instantiate Jenkins ephemeral in my development environment, RSS is  between 400-437M. 

I did some testing to ensure this difference wasn't just some early startup memory consumption. I don't see any early bubble in JVM memory usage. 

So... what could be different between 
`./test/extended/core.sh --ginkgo.focus="jenkins-plugin test connection test" --ginkgo.v`
in my development environment and CI test run 758?

**Comment by bparees** [comment]
@danmcp @smarterclayton who can we pull in from the jdk team to look at the jenkins memory usage (native+heap)?  We seem to be exceeding what we'd expect (we can't get jenkins to run with stability in a 512meg container with a 300meg max heap size)

**Comment by danmcp** [comment]
@bparees What makes us think that something is wrong or unexpected with the JDK?  Do we have a heap dump?

**Comment by bparees** [comment]
@danmcp we're seeing large variances in what we assume is native memory usage in different environments.  We're still working on getting a heap dump but we assume the JVM is respecting the heap limit.  The question is why it's using so much native memory. (the JVM RSS is 511megs) and whether there's anything we can do to bring it down.

**Comment by danmcp** [comment]
Does -XX:NativeMemoryTracking=summary work with openjdk?

From: http://stackoverflow.com/questions/2756798/java-native-memory-usage

**Comment by gabemontero** [comment]
On Tue, Nov 8, 2016 at 12:40 PM, Dan McPherson <notifications@github.com>
wrote:

> Does -XX:NativeMemoryTracking=summary work with openjdk?
>

According the `java -help`, there is only `java -X` with openjdk, and when
running that to get the list, `NativeMemoryTracking` is not listed.
However, I was able to follow the directions from the stackoverflow post
you referenced and `jcmd` produced the similar looking output.

**Comment by jupierce** [comment]
I've updated this PR to dump the following when the high memory condition is detected:
- Thread dump
- Object histogram

These won't directly tell us what is happening with native memory, but they could still guide our next steps.

I tried getting more detailed memory information using jmap, but the debuginfo package is stripped from our pod's JVM: https://bugzilla.redhat.com/show_bug.cgi?id=1010786 . Very limited information is therefore available.

I also tried jcmd, but received the message that native memory tracking is not enabled. 
```
# oc rsh jenkins-1-lfse8 jcmd 1 VM.native_memory
1:
Native memory tracking is not enabled
```

I assume we would need to alter our image to enable this.

**Comment by gabemontero** [comment]
On Tue, Nov 8, 2016 at 1:53 PM, Justin Pierce <notifications@github.com>
wrote:

> I've updated this PR to dump the following when the high memory condition
> is detected:
>
>    - Thread dump
>    - Object histogram
>
> These won't directly tell us what is happening with native memory, but
> they could still guide our next steps.
>
> I tried getting more detailed memory information using jmap, but the
> debuginfo package is stripped from our pod's JVM:
> https://bugzilla.redhat.com/show_bug.cgi?id=1010786 . Very limited
> information is therefore available.
>
> I also tried jcmd, but received the message that native memory tracking is
> not enabled.
>
> # oc rsh jenkins-1-lfse8 jcmd 1 VM.native_memory
> 1:
> Native memory tracking is not enabled
>
> I assume we would need to alter our image to enable this.
>
Yep we'd have to modify
https://github.com/openshift/jenkins/blob/master/1/contrib/s2i/run#L108 and
add the --XX:NativeMemoryTracking=summary option previously discussed and
then rebuild.

**Comment by jupierce** [comment]
[testextended][extended:core(image_ecosystem)]

**Comment by jupierce** [comment]
Test 769 (https://ci.openshift.redhat.com/jenkins/job/test_pr_origin_extended/769/consoleFull) included one Jenkins restart. Appears to have happened extremely early. Before initial metrics were gathered (i.e. < 10 seconds). 
Running again...

**Comment by jupierce** [comment]
[testextended][extended:core(image_ecosystem)]

**Comment by jupierce** [comment]
[testextended][extended:core(openshift pipeline plugin)]

**Comment by jupierce** [comment]
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin_extended/776/consoleFull
Shows memory growth and an OOM kill, but new logging statements were a moment too late to the party. Reduced the RSS threshold in response -- running tests again.
```
Unable to acquire Jenkins histogram: exit status 137
Nov  9 12:07:03.676: INFO: Running 'oc exec --config=/tmp/extended-test-jenkins-plugin-6kis9-r6b9i-user.kubeconfig --namespace=extended-test-jenkins-plugin-6kis9-r6b9i --namespace extended-test-jenkins-plugin-6kis9-r6b9i-jenkins jenkins-1-0sv14 -- jstack 1'
Nov  9 12:07:03.947: INFO: Error running &{/data/src/github.com/openshift/origin/_output/local/bin/linux/amd64/oc [oc exec --config=/tmp/extended-test-jenkins-plugin-6kis9-r6b9i-user.kubeconfig --namespace=extended-test-jenkins-plugin-6kis9-r6b9i --namespace extended-test-jenkins-plugin-6kis9-r6b9i-jenkins jenkins-1-0sv14 -- jstack 1] []   error: Internal error occurred: error executing command in container: container not found ("jenkins")
 error: Internal error occurred: error executing command in container: container not found ("jenkins")
 [] <nil> 0xc820a61660 exit status 1 <nil> true [0xc82002c9c8 0xc82002ca08 0xc82002ca08] [0xc82002c9c8 0xc82002ca08] [0xc82002c9d8 0xc82002ca00] [0xb187a0 0xb18900] 0xc820fdf920}:
error: Internal error occurred: error executing command in container: container not found ("jenkins")
Unable to acquire Jenkins thread dump: exit status 1
Nov  9 12:07:03.947: INFO: Running 'oc exec --config=/tmp/extended-test-jenkins-plugin-6kis9-r6b9i-user.kubeconfig --namespace=extended-test-jenkins-plugin-6kis9-r6b9i --namespace extended-test-jenkins-plugin-6kis9-r6b9i-jenkins jenkins-1-0sv14 -- cat /sys/fs/cgroup/memory/memory.stat'
Retrieving Jekins resource: "http://172.30.249.192:80/job/test-exec-dsl-job/lastBuild/buildNumber"
Error accessing resource: Unable to GET uri "http://172.30.249.192:80/job/test-exec-dsl-job/lastBuild/buildNumber": Get http://172.30.249.192:80/job/test-exec-dsl-job/lastBuild/buildNumber: dial tcp 172.30.249.192:80: getsockopt: connection refused
Nov  9 12:07:04.209: INFO: Error running &{/data/src/github.com/openshift/origin/_output/local/bin/linux/amd64/oc [oc exec --config=/tmp/extended-test-jenkins-plugin-6kis9-r6b9i-user.kubeconfig --namespace=extended-test-jenkins-plugin-6kis9-r6b9i --namespace extended-test-jenkins-plugin-6kis9-r6b9i-jenkins jenkins-1-0sv14 -- cat /sys/fs/cgroup/memory/memory.stat] []   error: Internal error occurred: error executing command in container: container not found ("jenkins")
 error: Internal error occurred: error executing command in container: container not found ("jenkins")
 [] <nil> 0xc821248000 exit status 1 <nil> true [0xc82002ca18 0xc82002ca40 0xc82002ca40] [0xc82002ca18 0xc82002ca40] [0xc82002ca20 0xc82002ca38] [0xb187a0 0xb18900] 0xc820fdfbc0}:
error: Internal error occurred: error executing command in container: container not found ("jenkins")

```

**Comment by jupierce** [comment]
[testextended][extended:core(openshift pipeline plugin)]

**Comment by jupierce** [comment]
[testextended][extended:core(openshift pipeline plugin)]

**Comment by jupierce** [comment]
Caught one:
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin_extended/779/console
Permanent: https://gist.github.com/jupierce/6fc451b276fc67193ed85c346cf9cc00

Analysis (you can search for the timestamps in bold to find the corresponding entries):

**2016-11-09 15:54:00.940790038 -0500 EST**
- Jenkins instance starts up at 476M RSS (-Xmx307m). This leaves a **minimum** native allocation of 169M.

**2016-11-09 15:54:20.94077847 -0500 EST**
- First time high memory utilization dump triggers
- RSS has  increased to 485M (9M increase in 20 seconds)
- Linux Cache: 26M
- Total memory: 511M
- Combined size of all heap objects 200M
- Thread count: 51

**2016-11-09 15:54:30.940791734 -0500 EST**
- Last *full* histogram collection before failures begin
- RSS has decreased to 484M
- Linux cache: 16M
- Total memory usage: 500M
- Combined size of all heap objects 205M
- Thread count: 51

**2016-11-09 15:54:40.941379079 -0500 EST**
- RSS/Cache/Total: 485/14/499M
- Unable to acquire Jenkins histogram: exit status 137

**2016-11-09 15:55:40.941167617 -0500 EST**
- New Jenkins instance now running (RSS=297M)

**Comment by bparees** [comment]
so if i'm understanding this, we're well within our max heap, but we're using massive amounts(200-300megs) of native memory in the jvm process?

**Comment by jupierce** [comment]
- Without knowing the amount of unused heap space, I'm not sure we can guess at anything other than the minimum size of the non-heap memory -- which peaked in the sampling at 178M. I'll work on extracting this information tomorrow. 

- We can be somewhat confident that the heap wasn't pressured around the time of the OOM based on the histogram.

- I've determined there is a sizeable difference between the RSS memory footprint of the openshift/jenkins image being run. For example, Jenkins-v1.651.2  (the image out on the hack day cluster) consumes about 40-50M less RSS than Jenkins-v2.7.3 (this difference exists when the total allocated heap size is identical). Variables which might be affecting this: the Jenkins version, the JVM, or the plugin version. The Jenkins version is the most likely culprit and I will try to prove it tomorrow.

**Comment by jupierce** [comment]
__Data__
I've run three different configurations of Jenkins in a test harness designed to monitor data about the JVM's memory usage. All configurations run in a pod with a 512M cgroup memory limit.

Configuration A: Heap=307M  (60% of total memory)    StackSize=1M
Configuration B: Heap=256M  (50% of total memory)    StackSize=1M
Configuration C: Heap=256M  (50% of total memory)    StackSize=500K

As part of each run, the same process was followed:
1. start the pod
2. begin monitoring it as soon as it reaches a steady memory state
3. induce a load on it (manually & programmatically logging in, navigating UI, creating jobs, etc)
4. stop inducing load and allow it to regain a steady state

Step (3) is the critical juncture of the test. Empirically, the **first** load on the system after startup creates the most predictable and largest spike in memory usage.  I believe it is this initial spike which occasionally triggers the OOM condition -- resulting in the extended test failures.

The spikes are in non-heap related areas of the JVM; particularly impacting the non-heap JVM metaspace content (metaspace is the new Java8 architecture for what used to be PermGen). Several short-lived, application specific, new threads are also created during this time to service some initial workload within Jenkins related to metrics. 

It is easy to visually verify that the heap is not growing or pressured during the test for any of the configurations:
screenshot from 2016-11-14 17-01-55

Moving on to the murkier non-heap memory utilization, the following graph shows the *total* committed JVM memory versus RSS for each configuration. To normalize the data, the values are shown as a percentage of total memory (e.g. 90% means 90% of 512M). The induced load on the configurations happens around or shortly before capture (6) in the diagram:
screenshot from 2016-11-14 17-16-41

Again, committed heap is not growing during this time period, so all fluctuations in committed memory are due to non-heap areas of the JVM. Looking at the committed memory lines proves how difficult it is to predict JVM memory management: Configuration B has a a significantly smaller heap size than configuration A, but it ultimately commits and uses more memory on average than A. Why it does this is a matter of speculation barring analysis of the JVM implementation (perhaps the JVM feels more confident it can meet future memory demands based on the smaller heap size and is greedier about allocating?).

Counter-intuitively, even though configuration B has more memory committed, I believe it would improve overall reliability in the constrained environment since it has pre-committed 100% of its heap and will not be able to dramatically increase it again under load (whereas configuration A has nearly 80M of uncommitted heap waiting to be allocated at any moment). What's more, configuration B will more aggressively garbage collect that committed heap to keep in check the amount of the committed memory that is actually in use.

Configuration C, which cuts the amount of stack each thread pre-allocates, does appear to effectively reduce the committed memory footprint of the JVM. It is the only configuration which never commits significantly more than 100% of available RAM.

__Summary__
I believe that reducing committed memory to less than 500M (leaving some memory for Linux cache) is crucial for stability. Configuration C nearly manages this and I believe it would dramatically decrease the overall odds of OOM conditions, but if the JVM wanted 100% of committed memory at the wrong instant, even the full 512M would not suffice.

To meet the target, I attempted constraining the JVM's metaspace, but this led to instability. So, to further reduce committed memory, another sizeable cut to the maximum heap size seems to be required. For example, a configuration with a 204M heap (40% of available memory) and 500K stack-size never committed more than 90% of available memory during testing. In this configuration, the heap is still around 30% unused. 

__Action__
I'd like to try tailoring the openshift/jenkins image run.sh script with a low-memory branch which reduces the thread size to 500K and heap to 40% of available memory. Workloads I was not able to generate during my testing may exhaust these resources, but, if they do, I'm hopeful we will see useful indications in the pod logs (e.g. memory messages from the JVM) instead of the difficult to detect OOM kills we presently experience.

**Comment by smarterclayton** [comment]
Are you running in 32bit or 64bit mode?

On Nov 15, 2016, at 10:20 AM, Justin Pierce <notifications@github.com>
wrote:

*Data*
I've run three different configurations of Jenkins in a test harness
designed to monitor data about the JVM's memory usage. All configurations
run in a pod with a 512M cgroup memory limit.

Configuration A: Heap=307M (60% of total memory) StackSize=1M
Configuration B: Heap=256M (50% of total memory) StackSize=1M
Configuration C: Heap=256M (50% of total memory) StackSize=500K

As part of each run, the same process was followed:
1. start the pod
2. begin monitoring it as soon as it reaches a steady memory state
3. induce a load on it (manually & programmatically logging in, navigating
UI, creating jobs, etc)
4. stop inducing load and allow it to regain a steady state

Step (3) is the critical juncture of the test. Empirically, the *first*
load on the system after startup creates the most predictable and largest
spike in memory usage. I believe it is this initial spike which
occasionally triggers the OOM condition -- resulting in the extended test
failures.

The spikes are in non-heap related areas of the JVM; particularly impacting
the non-heap JVM metaspace content (metaspace is the new Java8 architecture
for what used to be PermGen). Several short-lived, application specific,
new threads are also created during this time to service some initial
workload within Jenkins related to metrics.

It is easy to visually verify that the heap is not growing or pressured
during the test for any of the configurations:
[image: screenshot from 2016-11-14 17-01-55]
<https://cloud.githubusercontent.com/assets/19783215/20284710/4b018c30-aa8c-11e6-8b71-345b8db86757.png>

Moving on to the murkier non-heap memory utilization, the following graph
shows the *total* committed JVM memory versus RSS for each configuration.
To normalize the data, the values are shown as a percentage of total memory
(e.g. 90% means 90% of 512M). The induced load on the configurations
happens around or shortly before capture (6) in the diagram:
[image: screenshot from 2016-11-14 17-16-41]
<https://cloud.githubusercontent.com/assets/19783215/20285122/37137934-aa8e-11e6-87cf-f09fe70934c8.png>

Again, committed heap is not growing during this time period, so all
fluctuations in committed memory are due to non-heap areas of the JVM.
Looking at the committed memory lines proves how difficult it is to predict
JVM memory management: Configuration B has a a significantly smaller heap
size than configuration A, but it ultimately commits and uses more memory
on average than A. Why it does this is a matter of speculation barring
analysis of the JVM implementation (perhaps the JVM feels more confident it
can meet future memory demands based on the smaller heap size and is
greedier about allocating?).

Counter-intuitively, even though configuration B has more memory committed,
I believe it would improve overall reliability in the constrained
environment since it has pre-committed 100% of its heap and will not be
able to dramatically increase it again under load (whereas configuration A
has nearly 80M of uncommitted heap waiting to be allocated at any moment).
What's more, configuration B will more aggressively garbage collect that
committed heap to keep in check the amount of the committed memory that is
actually in use.

Configuration C, which cuts the amount of stack each thread pre-allocates,
does appear to effectively reduce the committed memory footprint of the
JVM. It is the only configuration which never commits significantly more
than 100% of available RAM.

*Summary*
I believe that reducing committed memory to less than 500M (leaving some
memory for Linux cache) is crucial for stability. Configuration C nearly
manages this and I believe it would dramatically decrease the overall odds
of OOM conditions, but if the JVM wanted 100% of committed memory at the
wrong instant, even the full 512M would not suffice.

To meet the target, I attempted constraining the JVM's metaspace, but this
led to instability. So, to further reduce committed memory, another
sizeable cut to the maximum heap size seems to be required. For example, a
configuration with a 204M heap (40% of available memory) and 500K
stack-size never committed more than 90% of available memory during
testing. In this configuration, the heap is still around 30% unused.

*Action*
I'd like to try tailoring the openshift/jenkins image run.sh script with a
low-memory branch which reduces the thread size to 500K and heap to 40% of
available memory. Workloads I was not able to generate during my testing
may exhaust these resources, but, if they do, I'm hopeful we will see
useful indications in the pod logs (e.g. memory messages from the JVM)
instead of the difficult to detect OOM kills we presently experience.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/11795#issuecomment-260669455>,
or mute
the thread

**Comment by jupierce** [comment]
@smarterclayton All of this data was gathered from 64b openjdk 1.8 JVM instances.

**Comment by bparees** [comment]
@jupierce ok, i think i'm sold on the "low memory config" path.

**Comment by smarterclayton** [comment]
Supposedly using 32bit jvms can halve small memory use.  Not sure that's
great if the jenkins instance scales past 3GB, but if you are fixed to 512
you should be in 32 bit mode.

**Comment by bparees** [comment]
> Supposedly using 32bit jvms can halve small memory use. 

i'm pretty sure compressed pointers did away with a large part of that difference.

**Comment by smarterclayton** [comment]
See
https://blog.buoyant.io/2016/06/17/small-memory-jvm-techniques-for-microservice-sidecars/

It's the base JVM libraries, not the pointers.

**Comment by jupierce** [comment]
[testextended][extended:core(openshift pipeline plugin)]

**Comment by jupierce** [comment]
[testextended][extended:core(openshift pipeline plugin)]

**Comment by jupierce** [comment]
[testextended][extended:core(openshift pipeline plugin)]

**Comment by bparees** [comment]
seems fixed?

**Comment by jupierce** [comment]
@bparees I've been using this PR to drive multiple runs of the extended tests -- they are passing reliably with the low memory profile change in openshift/jenkins#197 . The logging introduced by this PR also indicates the low memory profile is having the desired impact.

My last push addresses @gabemontero's feedback -- minor refactor in how I find the Jenkins pod. ptal at DumpDeploymentLogs & findJenkinsPod. If you are satisfied with the change and tests pass, I am ready to merge.

**Comment by gabemontero** [comment]
@jupierce - quick question, and I apologize if I missed it in the earlier discourse, but if -Xss is *not* set, what is the default java thread stack size?

**Comment by jupierce** [comment]
@gabemontero - 1M is the current default for stack size.

**Comment by gabemontero** [comment]
cool - thx @jupierce

**Comment by jupierce** [comment]
[testextended][extended:core(builds)]

**Comment by jupierce** [comment]
[testextended][extended:core(openshift pipeline plugin)][extended:core(builds)]

**Comment by jupierce** [comment]
Jenkins test passed with changes for @gabemontero . Increased test scope since other extended tests use DumpDeploymentConfig.

**Comment by bparees** [comment]
[merge]

**Comment by smarterclayton** [comment]
I'd like to know how possible running a 32bit jvm is when limit <2Gb in the same image and what the trade offs are.

**Comment by jupierce** [comment]
@smarterclayton I will gather statistics for a 32b JVM under the same conditions.

**Comment by jupierce** [comment]
@smarterclayton @bparees @gabemontero 
Please see the update with a 32b JVM. The x86jvm lines have the same configuration as "A" (307M heap and 1M thread stack size; i.e. no tailoring for low memory). 

screenshot from 2016-11-18 12-03-44

There is a substantial reduction in the committed memory that might obviate any tinkering with JVM memory parameters. One tradeoff here is that centos7 does not appear to build an x86 openjdk -- I had to reach back to centos 6.5 RPMs to execute this test.

**Pull request merged**